packages/flutter_tools/lib/src/android/migrations/gradle_java_version_conflict_migration.dart

```diff
@@ -0,0 +1,72 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import '../../base/file_system.dart';
+import '../../base/project_migrator.dart';
+import '../../base/version.dart';
+import '../../project.dart';
+import '../android_studio.dart';
+
+final Version _androidStudioFlamingo = Version(2022, 2, 0);
+final RegExp _gradleVersionMatch = RegExp(
+  r'\s*distributionUrl=https\\://services\.gradle\.org/distributions/gradle-((?:\d|\.)+)-(?:all|bin)\.zip');
+final Version _lowestSupportedGradleVersion = Version(7, 3, 0);
+const String _newGradleVersionString = r'7.4';
+
+/// Migrate to a newer version of gradle when the existing does not support
+/// the version of Java provided by the detected Android Studio version.
+///
+/// For more info see the Gradle-Java compatibility matrix:
+/// https://docs.gradle.org/current/userguide/compatibility.html
+class GradleJavaVersionConflictMigration extends ProjectMigrator {
+  GradleJavaVersionConflictMigration(
+      AndroidProject project,
+      super.logger,
+      AndroidStudio? androidStudio,
+      ) : _androidStudio = androidStudio,
+          _gradleWrapperPropertiesFile = project.hostAppGradleRoot
+      .childDirectory('gradle').childDirectory('wrapper').childFile('gradle-wrapper.properties');
+  final File _gradleWrapperPropertiesFile;
+  final AndroidStudio? _androidStudio;
+
+  @override
+  void migrate() {
+    if (!_gradleWrapperPropertiesFile.existsSync()) {
+      logger.printTrace('gradle-wrapper.properties not found, skipping gradle version compatibility check.');
+      return;
+    }
+
+    if (_androidStudio == null) {
+      logger.printTrace('Android Studio version could not be detected, '
+          'skipping gradle version compatibility check.');
+      return;
+    } else if (_androidStudio!.version < _androidStudioFlamingo) {
```

Comment:
I think ideally this would be based directly on what JDK the tool is planning to use for running Gradle. When we're successfully finding an Android Studio install, and successfully locating its bundled JDK, that's the JDK we use; but when we don't, we find another JDK, and the build will be just as broken if we have an old Gradle and a JDK 19 the user set in JAVA_HOME as if we have an old Gradle and a JDK 17 the user got from Android Studio Flamingo.

It'd be reasonable to wait for #124152 first, though, to clean up the concept of "what JDK are we using". Focusing on the Android Studio version is good enough to cover the impending upgrade that's urgent right now.